### PR TITLE
Improve error messages

### LINKED ISSUE #48: Improve error messages
Update error message from github actions to include a direct link to where to get additional details.

### PR DESCRIPTION
**Pull request by baijum**
- Link to the job failure page
- Include a summary of errors
- Username is not part of the OWNERS
- chart name and version should be the same in the directory structure, OWNERS file, Chart.yaml, report.yaml
- Digest mismatch message

Fixes #48

### CODE REVIEW

File: scripts/src/chartprreview/chartprreview.py [hunk 1]
```diff
@@ -14,27 +14,34 @@
...
     pattern = re.compile(r"charts/(\w+)/([\w-]+)/([\w-]+)/([\w\.]+)/.*")
-    count = 0
     for f in r.json():
         m = pattern.match(f["filename"])
         if m:
             category, organization, chart, version = m.groups()
             return category, organization, chart, version
 
-    return "", "", "", ""
+    msg = "[ERROR] The files modified are not part of any chart"
```
Comment (mmulholla) [inline scripts/src/chartprreview/chartprreview.py:35]: Maybe "One or more files included in the pull request are not part of the chart"

File: scripts/src/chartprreview/chartprreview.py [hunk 2]
```diff
@@ -14,27 +14,34 @@
...
+    sys.exit(1)
 
-
-def verify_user(username, category, organization, chart):
+def verify_user(directory, username, category, organization, chart):
     data = open(os.path.join("charts", category, organization, chart, "OWNERS")).read()
     out = yaml.load(data, Loader=Loader)
     if username not in [x['githubUsername'] for x in out['users']]:
-        print("User doesn't exist in list of owners:", username)
+        msg = f"[ERROR] User doesn't exist in list of owners: {username}"
```
Comment (mmulholla) [inline scripts/src/chartprreview/chartprreview.py:43]: Maybe "${username} who submitted the pull request is not included in the list of owners"

File: scripts/src/chartprreview/chartprreview.py [hunk 3]
```diff
@@ -43,16 +50,18 @@ def check_owners_file_against_directory_structure(username, category, organizati
     vendor_label = out["vendor"]["label"]
     chart_name = out["chart"]["name"]
     error_exit = False
+    msgs = []
     if organization != vendor_label:
         error_exit = True
-        print("vendor/label in OWNERS file doesn't match the directory structure.")
+        msg.append("[ERROR] vendor/label in OWNERS file doesn't match the directory structure")
```
Comment (mmulholla) [inline scripts/src/chartprreview/chartprreview.py:56]: Maybe  vendor/label in OWNERS file (${vendor_label} doesn't match the directory structure (${organization})

or other way if including the expected and sbmitted values.

Same for chart name message below.

File: scripts/src/chartprreview/chartprreview.py [hunk 4]
```diff
@@ -79,41 +88,50 @@ def match_checksum(category, organization, chart, version):
...
+def match_name_and_version(directory, category, organization, chart, version):
     submitted_report_path = os.path.join("charts", category, organization, chart, version, "report.yaml")
     if os.path.exists(submitted_report_path):
         submitted_report = yaml.load(open(submitted_report_path), Loader=Loader)
         submitted_report_chart_name = submitted_report["metadata"]["chart"]["name"]
         submitted_report_chart_version = submitted_report["metadata"]["chart"]["version"]
 
         if submitted_report_chart_name != chart:
-            print("Chart name is not matching against the value in the directory structure:", submitted_report_chart_name, "vs.", chart)
+            msg = f"[ERROR] Chart name is not matching against the value in the directory structure: {submitted_report_chart_name} vs. {chart}"
```
Comment (mmulholla) [inline scripts/src/chartprreview/chartprreview.py:128]: repeat of previous message? Make them the same.

**Comment by mmulholla** [comment]
Generally make messages as clear as possible, including failed and expected values wherever possible. Make it easy for the user to see exactly where the problem lies.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -11,6 +11,13 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v2
 
+      - name: save check run id
+        run: |
+          CHECK_SUITE_URL=$(curl -s -H "authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" https://api.github.com/repos/${GITHUB_REPOSITORY}/actions/runs/${GITHUB_RUN_ID} | jq -r '.check_suite_url')
+          CHECK_RUN_ID=$(curl -s -H "authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" -H "Accept: application/vnd.github.antiope-preview+json" $CHECK_SUITE_URL/check-runs | jq '.check_runs[] | select(.name=="build-and-verify") | .id ')
+          mkdir -p ./pr
+          echo $CHECK_RUN_ID > ./pr/run-id
```
Comment (pedjak) [inline .github/workflows/build.yml:19]: where is `run-id` used, i.e. read?

File: scripts/src/chartprreview/chartprreview.py (hunk 2 above)
Comment (pedjak) [reply scripts/src/chartprreview/chartprreview.py:43]: Perhaps "{usernmae} is not allowed to submit the chart on behalf of {partner}"

File: scripts/src/chartprreview/chartprreview.py (hunk 3 above)
Comment (pedjak) [reply scripts/src/chartprreview/chartprreview.py:56]: We should tell use how this structure should look like.

File: .github/workflows/build.yml (hunk 5 above)
Comment (baijum) [reply .github/workflows/build.yml:19]: I have sent another PR #77 
It's used there to link to the error log console.
https://github.com/openshift-helm-charts/repo/pull/77/files#diff-e29b824e0807b8c6fe5c1e8538353b8a0dd88850281bce900dfb90bd9f73520dR6-R11

File: scripts/src/chartprreview/chartprreview.py (hunk 1 above)
Comment (baijum) [reply scripts/src/chartprreview/chartprreview.py]: Done. 229bd4e5c3b304f1c759c7b5b8f51971defae155

File: scripts/src/chartprreview/chartprreview.py (hunk 2 above)
Comment (baijum) [reply scripts/src/chartprreview/chartprreview.py]: I have taken @pedjak 's suggestion here: 229bd4e5c3b304f1c759c7b5b8f51971defae155

File: scripts/src/chartprreview/chartprreview.py (hunk 3 above)
Comment (baijum) [reply scripts/src/chartprreview/chartprreview.py]: Done. 229bd4e5c3b304f1c759c7b5b8f51971defae155

File: scripts/src/chartprreview/chartprreview.py (hunk 4 above)
Comment (baijum) [reply scripts/src/chartprreview/chartprreview.py]: Done. 229bd4e5c3b304f1c759c7b5b8f51971defae155

**Comment by baijum** [comment]
@mmulholla @pedjak I have addressed all the comments. Please take a look again!

File: .github/workflows/build.yml (hunk 5 above)
Comment (pedjak) [reply .github/workflows/build.yml:19]: If it used there, then please remove it from this one and move it overthere.

Comment (baijum) [reply .github/workflows/build.yml:19]: This job doesn't have access to create comments. So, all the required information is captured in the PR artifact. Later the other job create comments based on the PR artifact.

Comment (pedjak) [reply .github/workflows/build.yml:19]: please keep this PR focused., introduced step "save check run id" is not consumed within this PR. Hence, if it is consumed in #77 - then add "save check run id" within that PR.

Comment (baijum) [reply .github/workflows/build.yml:19]: This PR is targeted at `dev` branch whereas the other PR is targeted at `main` branch.

**Comment by baijum** [comment]
@pedjak @mmulholla I have pushed one more commit. Please check.
5561ddb59eb0f8cc3eda7193187ecf54632684a2

File: .github/workflows/build.yml (hunk 5 above)
Comment (pedjak) [reply .github/workflows/build.yml:19]: Does this mean that `run-id` is not consumed in `dev` branch at all? If so, why adding it here?

Comment (baijum) [reply .github/workflows/build.yml:19]: The `run-id` is the run ID of the current job. For example, in this PR, as of now this in the URL for `Build and Verify`: https://github.com/openshift-helm-charts/repo/pull/69/checks?check_run_id=2499764781
When the job is triggered again, the `run-id` will change.

This information should be captured while this job is running. Then only we can add it as a PR comment from the other job.

**Comment by mmulholla** [comment]
LGTM once Predrag is OK.

**Comment by baijum** [comment]
> LGTM once Predrag is OK.

Thanks @mmulholla ! I will wait for approval from @pedjak before merging.

File: .github/workflows/build.yml (hunk 5 above)
Comment (pedjak) [reply .github/workflows/build.yml:19]: ok, so iirc, the job that consumes this `run-id` does not even exist on `dev` branch, hence this change needs to get here.

Comment (pedjak) [inline .github/workflows/build.yml:19]: can we call this file `build-verify-check-run-id` to reflect what is the value is?

Comment (baijum) [reply .github/workflows/build.yml]: Done. 6bb4c0d8fd39a892f391f3c7eead832a1985f031

**Comment by baijum** [comment]
@pedjak @mmulholla I have added one more commit: 8099c4aae15a862c662cdda5c3b66e649368e3f6

Please note, that's _not_ an error, so the PR can still be merged. I think a warning for 4XX and 5XX URL status would be helpful if they make a genuine mistake. Otherwise, we can assume they know what they are doing.

I know the PR will be automatically merged, and they will miss this warning. I think we can make a recommendation from our documentation always to read the messages in the PR even if it gets merged.

**Comment by pedjak** [comment]
> I know the PR will be automatically merged, and they will miss this warning. I think we can make a recommendation from our documentation always to read the messages in the PR even if it gets merged.

Well, the notification on new comments on closed PR will be sent at least to PR author, so I think we are good here.

Please squash your commits before merging this PR.

**Pull request merged**